// leaflet-routing-machine/src/L.Routing.Localization.js
(function() {
	'use strict';
	L.Routing = L.Routing || {};

	L.Routing.Localization = {
		'en': {
			directions: {
				N: 'north',
				NE: 'northeast',
				E: 'east',
				SE: 'southeast',
				S: 'south',
				SW: 'southwest',
				W: 'west',
				NW: 'northwest'
			},
			instructions: {
                'turn' : {
                  'uturn': ['Turn around', ' and continue on {road}'],
                  'sharp right': ['Turn sharp right', ' onto {road}'],
                  'right': ['Turn right', ' onto {road}'],
                  'slight right': ['Turn slight right', ' onto {road}'],
                  'straight': ['Go straight', ' onto {road}'],
                  'slight left': ['Turn slight left', ' onto {road}'],
                  'left': ['Turn left', ' onto {road}'],
                  'sharp left': ['Turn sharp left', ' onto {road}']
                },
				// instruction, postfix if the road is named
                'depart':{
                  'uturn': ['Leave', ' towards {road}'],
                  'sharp right': ['Leave', ' towards {road}'],
                  'right': ['Leave', ' towards {road}'],
                  'slight right': ['Leave', ' towards {road}'],
                  'straight': ['Leave', ' towards {road}'],
                  'slight left': ['Leave', ' towards {road}'],
                  'left': ['Leave', ' towards {road}'],
                  'sharp left': ['Leave', ' towards {road}'],
                },
				'arrive':{
                  'uturn': ['You have reached your destination'],
                  'sharp right': ['You have reached your destination'],
                  'right': ['You have reached your destination, your destination is on the right'],
                  'slight right': ['You have reached your destination'],
                  'straight': ['You have reached your destination, your destination is ahead of you'],
                  'slight left': ['You have reached your destination'],
                  'left': ['You have reached your destination, your destination is on the left'],
                  'sharp left': ['You have reached your destination'],
                },
				'waypoint':{
                  'uturn': ['You have reached your intermediate destination'],
                  'sharp right': ['You have reached your intermediate destination'],
                  'right': ['You have reached your intermediate destination, your destination is on the right'],
                  'slight right': ['You have reached your intermediate destination'],
                  'straight': ['You have reached your intermediate destination, your destination is ahead of you'],
                  'slight left': ['You have reached your intermediate destination'],
                  'left': ['You have reached your intermediate destination, your destination is on the left'],
                  'sharp left': ['You have reached your intermediate destination'],
                },
                'continue':{
                  'uturn': ['Turn around'],
                  'sharp right': ['Turn sharp right', ' to stay on {road}'],
                  'right': ['Turn right', ' to stay on {road}'],
                  'slight right': ['Turn slight right', ' to stay on {road}'],
                  'straight': ['Go straight', ' to stay on {road}'],
                  'slight left': ['Turn slight left', ' to stay on {road}'],
                  'left': ['Turn left', ' to stay on {road}'],
                  'sharp left': ['Turn sharp left', ' to stay on {road}']
                },
                'new name':{
                  'uturg': ['This should never hapen'],
                  'sharp right': ['The road takes a sharp turn to the right', ' and becomes {road}'],
                  'right': ['The road turns right', ' and becomes {road}'],
                  'slight right': ['The road takes a slight turn to the right', ' and becomes {road}'],
                  'straight': ['The road name changes', ' to {road}'],
                  'slight left': ['The road takes a slight turn to the left', ' and becomes {road}'],
                  'left': ['The road turns left', ' and becomes {road}'],
                  'sharp left': ['The road takes a sharp turn to the left', ' and becomes {road}']
                },
                'merge':{
                  'uturn': ['Merge', ' onto {road}'],
                  'sharp right': ['Merge', ' onto {road}'],
                  'right': ['Merge', ' onto {road}'],
                  'slight right': ['Merge', ' onto {road}'],
                  'straight': ['Merge', ' onto {road}'],
                  'slight left': ['Merge', ' onto {road}'],
                  'left': ['Merge', ' onto {road}'],
                  'sharp left': ['Merge', ' onto {road}'],
                },
                'ramp':{
                  'uturn': ['This should never happen'],
                  'sharp right': ['Make a sharp right and take the ramp', ' ({road})', ' towards {destination}'],
                  'right': ['Turn right and take the ramp', ' ({road})', ' towards {destination}'],
                  'slight right': ['Make a slight right and take the ramp', ' ({road})', ' towards {destination}'],
                  'straight': ['Go straight onto the ramp', ' ({road})', ' towards {destination}'],
                  'slight left': ['Make a slight right and take the ramp', ' ({road})', ' towards {destination}'],
                  'left': ['Turn left and take the ramp', ' ({road})', ' towards {destination}'],
                  'sharp left': ['Make a sharp left and take the ramp', ' ({road})', ' towards {destination}'],
                },
                'fork':{
                  'uturn': ['This should never happen'],
                  'sharp right': ['This should never happen'],
                  'right': ['At the fork, keep right', ' and continue on {road}'],
                  'slight right': ['This should never happen'],
                  'straight': ['This should never happen'],
                  'slight left': ['This should never happen'],
                  'left': ['At the fork, keep left', ' and continue on {road}'],
                  'sharp left': ['This should never happen'],
                },
                'end of road':{
                  'uturn': ['This should never happen'],
                  'sharp right': ['This should never happen'],
                  'right': ['At the end of the road, turn right', ' onto {road}'],
                  'slight right': ['This should never happen'],
                  'straight': ['This should never happen'],
                  'slight left': ['This should never happen'],
                  'left': ['At the end of the road, turn left', ' onto {road}'],
                  'sharp left': ['This should never happen'],
                },
                'roundabout':{
                  'uturn': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'sharp right': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'right': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'slight right': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'straight': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'slight left': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'left': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                  'sharp left': ['At the roundabout, take the {exitStr} exit', ' onto {road}'],
                },
			},
			formatOrder: function(n) {
				var i = n % 10 - 1,
				suffix = ['st', 'nd', 'rd'];

				return suffix[i] ? n + suffix[i] : n + 'th';
			},
			ui: {
				startPlaceholder: 'Start',
				viaPlaceholder: 'Via {viaNumber}',
				endPlaceholder: 'End'
			}
		},
	};

	module.exports = L.Routing;
})();
